### ISSUE TITLE
Raft table metadata behavior

### ISSUE THREAD

**Issue by timmaxw** [rethinkdb/rethinkdb#3594]
Currently, RethinkDB stores table metadata in the semilattices. This means that every server has a copy of the metadata, including a complete list of the names of all of the tables that exist. However, we are planning to change this behavior as part of the Raft project; the table metadata will instead be managed by a Raft instance for that table.

One major side effect is that under the new system, a server will not know that a table exists unless it can contact at least one of the servers hosting the table. For example, suppose that we have servers `A`, `B`, `C`, and `D`, where `A`, `B`, and `C` are hosting table `foo`. Under normal conditions, `r.table("foo")` will work on any server. But if server `D` gets isolated from servers `A`, `B`, and `C`, then `r.table("foo")` will return `Table "foo" does not exist.` instead of `Cannot reach master for table "foo".`. `foo` will also disappear from `table_list()`, `rethinkdb.table_config`, etc. This is a significant change from the pre-Raft behavior, in which every server knows about every table even when not directly connected.

Another side effect is that the table metadata cannot be changed unless a quorum of the servers hosting the table are available. For example, suppose server `A` is isolated in the example above. In the pre-Raft behavior, the user can ask server `A` to unilaterally change the table's configuration, and the change will propagate to the other servers when the netsplit is resolved. In the proposed Raft implementation, the configuration cannot be changed unless two of `A`, `B`, and `C` are available (although we will also provide the user with an "unsafe override" mechanism that works even if only one is available).

Comments?

**Comment by Tryneus** [comment]
I think the first side effect is a fairly serious problem.  I'm positive some of our users have scripts that will run something along the lines of:

```py
if table_name not in r.table_list().run(c):
  r.table_create('foo').run(c)
# Do work on the table 'foo'
```

I see two options for these users under this changed behavior:

1. They have to be very aware of which servers are hosting the tables and which servers their scripts are connecting to.
2. They have to change their provisioning strategy to create all tables ahead of time and recognize that `Table "foo" does not exist.` means that there is an availability issue.

Both of these options seem counter to what we're trying to achieve, but perhaps there is some other way to solve this.

I don't think the second side-effect is a huge deal, the biggest problem would be if a user loses some server data directories, resulting in a table that can never be reconfigured (though we should absolutely fix #3520 so they could possibly copy the table in this case).

**Comment by coffeemug** [comment]
> One major side effect is that under the new system, a server will not know that a table exists unless it can contact at least one of the servers hosting the table.

I think this might be a problem. May be we could transmit an eventually consistent list of table names to every node in the cluster (possibly by duplicating it into the semilattices)? 

> Another side effect is that the table metadata cannot be changed unless a quorum of the servers hosting the table are available.

This is a reasonable tradeoff, I wouldn't worry about this side effect.

**Comment by timmaxw** [comment]
After a lot of in-person discussion, we decided on the following: Each server will store a non-authoritative list of all the tables it knows about. This list will be stored and transmitted in approximately the same way as the current semilattices.

As part of the same discussion, we decided to do the same thing for servers. So we changed our minds about the principle that "if you disconnect a server from the cluster, you shouldn't get an error message if it didn't have any responsibilities".

**Comment by larkost** [comment]
Jut to amplify on the non-authoritative part: the only place that this would be used is for listing and error messages. So @Tryneus's example would still do the right thing and not try and generate a new table, and server admins's will not panic when we give them a false "this table does not exist" message (rather they will panic that that server can't connect to that table, which is the correct thing to do).

**Comment by deontologician** [comment]
As someone who's completely naive to all this, there's obviously some point to having raft over the semilattices. Why are they not good enough for an authoritative table list, but it's ok for non-authoritative?

**Comment by thelinuxlich** [comment]
A noob question and a bit off-topic: does this affect the amount of memory RethinkDB eats for each table?

**Comment by danielmewes** [comment]
@thelinuxlich That's mostly independent of this, except that a table is not going to consume as much memory on servers which aren't a replica for the table as they do now.

**Comment by coffeemug** [comment]
> As part of the same discussion, we decided to do the same thing for servers. So we changed our minds about the principle that "if you disconnect a server from the cluster, you shouldn't get an error message if it didn't have any responsibilities".

I think we should this discuss this a bit more. I could see arguments for this and against this, and it's not clear to me which is better. @VeXocide @larkost -- @timmaxw told me you have strong opinions on this -- would you mind sharing them here?

**Comment by timmaxw** [comment]
@deontologician: I haven't thought through the details yet, but here's a preliminary answer: Raft is necessary for auto-failover because it gives stronger consistency guarantees than the semilattices. But it's not strictly necessary to store the table config in Raft; we could hypothetically keep the table config in the semilattices and just use Raft for auto-failover-related state. Also, the place where we'll be storing non-authoritative data might not act exactly like the semilattices.

**Comment by larkost** [comment]
@coffeemug: My focus is mostly on the experience of the sys-admin and with that in mind the defects of the original proposal are:
1. The admin getting an incorrect "no such table" when connecting to an net-isolated computer that was not a shareholder for that table. This would lead to admins incorrectly panicking that "RethinkDB lost my data!". Adding more verbiage about "or could be not visible" feels like low-qulity punting.
2. The problem [nicely written up](https://github.com/rethinkdb/rethinkdb/issues/3594#issuecomment-70338179) by @Tryneus that would cause duplicate tables.
3. If I were a sysadmin of a 100-computer RethinkDB cluster I would need to know when one of my machines was not connected to the cluster. Having RethinkDB notice this and present that information on the admin pages makes for a really nice system.

**Comment by timmaxw** [comment]
> I don't think the second side-effect is a huge deal, the biggest problem would be if a user loses some server data directories, resulting in a table that can never be reconfigured

Actually, we're planning to provide an "escape hatch" in case this happens. The user can reconfigure the table in a "hard override" mode, which works as long as at least one replica is available, but which may result in limited data loss because the replica might not have the most up-to-date writes. So this probably isn't actually a problem.

**Comment by coffeemug** [comment]
I'm on board with maintaining a list of tables, but I think maintaining a list of servers is worth debating a bit.

> If I were a sysadmin of a 100-computer RethinkDB cluster I would need to know when one of my machines was not connected to the cluster.

Admittedly I've never administered a large cluster of machines, but consider the opposite argument. I have a RethinkDB cluster, and I opportunistically provision extra servers in case I later need to scale up some tables. None of the extra servers actually have anything on them, this is just an opportunistic measure. At that scale, some of the hardware will sometimes fail. Perhaps some servers will drop off the face of the earth, and some automated software will provision more. These could also be VMs, or Docker containers, or whatever.

When a server like that dies, it doesn't impact my RethinkDB cluster in any way, and my automated software (say, CloudFormation on EC2) will notice it and spin up another server. In this context, having to take care of a RethinkDB issue is just a schlep for me -- a task I have to perform that doesn't help me, but that I have to think about. I think that on balance it would make my life slightly harder (although I could also see the opposing point of view).

**Comment by larkost** [comment]
@coffeemug: Having admined a 300+ computer group (not quite a cluster, but close) I absolutely wanted something to tell me that a computer had dropped off the system. Note that this is separate from the computer being down, in many cases the computer was running, but not in communication with or not running my service.

This does not mean that the system should stop working, or not self-heal. But I want:
1. The list of all of the current problems.
2. A simple panel that I can go to get status on all of my servers to reassure myself that all is green[1].
3. The full log of when things went bad and when they got fixed (how is secondary)

The idea that servers could come and go from my cluster without me being aware of it scares me. Making it easy for me to both add and remove servers is definitely a plus, but doing so magically gives me the willies.

[1] It is amazing how obsessed many managers get about having a dashboard with a bunch of green indicators. Us providing a one-stop location for both the "good" and "bad" items is a very attractive thing for admins.

**Comment by coffeemug** [comment]
Ok. Unless someone else has an opposite opinion, I remove my objection.

**Comment by deontologician** [comment]
@larkost: you adminned a very different group of computers than @coffeemug is talking about. Where the machines are mostly virtual, and having the cluster administration software spin up new ones if there's a hiccup is expected and normal. You'll want logs to see that something happened, but an individual virtual server being thrown out and a new one provisioned isn't always a big deal in cloud environments.

That being said, it doesn't sound like there's a one-size fits all answer. Both use cases seem realistic to me. Maybe if you're using heavy cloud automation for your cluster, you wouldn't mind changing a config to turn off server tracking. It seems like the common small cluster case is to care about your servers

**Comment by coffeemug** [comment]
We could add a `--no-server-tracking` optarg that would control this behavior... I could also ping some people who run large clusters and ask for their opinion on this.

**Comment by VeXocide** [comment]
I think it makes sense to ping some people who run large clusters to ask for their opinion. My opinion was primarily due to the large cluster use case, but I had only a single datapoint to rely on.

**Comment by coffeemug** [comment]
I will ping some people and report back.

**Comment by danielmewes** [comment]
@coffeemug Have you got around asking some people?

**Comment by coffeemug** [comment]
Yes, I did.

Pretty much everyone I talked to told me they consider servers/processes to be ephemeral  (that is, they agree with @deontologician and not with @larkost). People got used to EC2, VMs, and Docker and they basically learned that you cannot expect a node/process to stay up at scale. So they expect nodes/processes to fail, and use an existing set of tools to monitor failing nodes and spin up new ones (EC2, for instance, has a bunch of APIs to do this; CoreOS does too, etc.) People were pretty adamant that the world is moving in the direction of ephemeral services, and that they already have tools they prefer to deal with the problem.

So I think RethinkDB nodes that aren't responsible for any data should be transient -- a node should be able to enter and leave a cluster arbitrarily, and we should only report issues when a node that's responsible for some data leaves the cluster.

**Comment by danielmewes** [comment]
Thanks @coffeemug . So it sounds like we should keep an eventually consistent non-authorative list of existing tables on each server, but not a list of connected servers.

The remaining question then is how you get an outdated entry out of that table list. Some obvious options:
- have a "cleanup" command: `r.purge_table_list()`
- let entries in the table list expire after a while of not seeing any of the responsible servers (this alone wouldn't be sufficient though if you need to recreate a table immediately)
- keep some notion of removing a server from the cluster, which would remove all tables from the list for which this server was the last one involved. That would probably be somewhat awkward though without also maintaining a local server list.
- let users run `table_drop()` even if the raft cluster associated to the table cannot be reached at the moment. We'd have to handle the case where one of the servers originally responsible for the table comes back though.

**Comment by coffeemug** [comment]
> have a "cleanup" command: `r.purge_table_list()`

I would prefer not to do it if possible. This is leaking an implementation detail to the user. Same for the notion of removing a server from a cluster (it would be really nice to get away from that completely).

I think expiring the non-authoritative list would be ideal. We could add a `force` optarg to `table_create` to deal with the instantaneous recreation case if we want to. In case the cluster is all connected, setting the optarg to `true` would create a duplicate table (this way it doesn't feel like too big an implementation leak/transgression).

**Comment by danielmewes** [comment]
I'm not a big fan of time-based expiration. I think it's a little weird to have unreachable tables still be there at first and then quietly disappear after some time.

How about we use `table_drop` with a special optarg to expire an entry from the table cache? Something like this:
```
> r.table_drop("unreachable").run()
Error: No server for table `unreachable` is connected. To remove the table from the table list, please use the `forget_unreachable=true` optarg. Note that the table will reappear if one of its servers reconnects later.

> r.table_drop("unreachable", forget_unreachable=true).run()
{tables_dropped: 1}
```

Both a time-based expiration and the `forget_unreachable` flag open the question of how we coordinate the expiration process between servers in the cluster, if some servers are unavailable when the command is run. Tables could re-appear if previously disconnected servers reconnect. What's worse is that an unreachable table could sometimes appear and sometimes not appear in a `table_list()` depending on which server that query is run on. This becomes especially bad if the cluster connectivity is not transitive.
Time-based expiration is slightly more benign with respect to these issues because the response to a `table_list()` is eventually consistent assuming full cluster connectivity over a sufficiently long time.

More sophisticated coordination (say something akin to our semilattice metadata) sounds like an unreasonable amount of work to implement for this.


I think it would be ok if we had only time-based expiration and ignored all of these problems if the expiration time was relatively short (say 10 minutes). That would help us get over most server restarts etc. I wonder though if it's even worth the effort in that case.

At this point I'm almost leaning towards discouraging our users from using patterns like
```
if 'foo' not in r.table_list().run(c):
  r.table_create('foo').run(c)
# Do work on the table 'foo'
```
if they're running any slightly larger cluster and otherwise ignoring the problem (i.e. making tables disappear immediately and just adapting the error message on a missing table to mention the fact that the table might just be unreachable).
In fact this pattern is already unsafe in a cluster, since table creation is not atomic.

**Comment by larkost** [comment]
I still maintain that we should never have a "or might be unreachable" error message. That is just going to cause "RethinkDB lost my data" panic. I don't think that we should ever lose track of the table list on any server (within what we can control). If you lose all of the shard-holders for a table, then there should be a way for an admin to force the deletion of the table to fix the solution, and @danielmewes's sounds fine to me.

As to @danielmewes's thought about checking for a table before using it: that seems like an odd worry to me. With two exceptions (testing like we do, and temporary tables), table creation is almost always done by either a setup script or an admin manually doing it. Anyone who is constantly checking for table existence before working with it is probably doing something wrong, and we can't control for stupid.

**Comment by danielmewes** [comment]
Hmm maybe we should resurrect a little piece of semilattice metadata for non-authorative table-existence management.
That metadata would only store the table IDs and table names. It would be persisted and synchronized between connected servers.
When an unreachable table is purged from the cache through `forget_unreachable` (or potentially other mechanisms), we generate a deletion entry in the table cache semilattice.
At any time, the actual table information from the directory takes precedence over caches table information and will overwrite it eventually.

@timmaxw how much work would it be to bring back some semilattice metadata for this purpose? Could we re-use our existing old code for managing it?

**Comment by coffeemug** [comment]
I retract my previous opinion wrt time-based expiration of tables. I agree with @larkost and @danielmewes (I know there are some open questions remaining, but I agree that forgetting unreachable tables should be an admin operation).

**Comment by timmaxw** [comment]
This is in CR 2839. We don't support any type of expiration of tables, aside from deleting them. If we have problems with users accidentally merging unrelated clusters, we should add protections against that.

**Comment by coffeemug** [comment]
:clap: :+1: :tada:

**Comment by timmaxw** [comment]
Merged into `raft` in 5d6e694db663603bec7cb897d54336ed6a7fb4dd.
